### PR TITLE
Second approach for #4295

### PR DESCRIPTION
**Pull request by ghost**
See `git diff leakage leak2`

I'm not sure how `reset()` exactly works. It seems, this is a better approach (I mean, any format must call `release_clobj` inside `create_clobj`).

### CODE REVIEW

**Comment by ghost** [comment]
This PR fixes all leaks in my formats.

**Review by magnumripper — APPROVED** [review]

File: src/opencl_rawsha256_fmt_plug.c
Comment (magnumripper) [inline src/opencl_rawsha256_fmt_plug.c:169]: The above probably needs to be added to all other (not yours) formats as well, right?

Comment (ghost) [reply src/opencl_rawsha256_fmt_plug.c:169]: Yes, using bash-fu (if possible). I use a fancy IDE.

File: src/opencl_rawsha256_fmt_plug.c
Comment (solardiz) [inline src/opencl_rawsha256_fmt_plug.c:304]: Extra empty line before the closing curly brace. Drop it.

File: src/opencl_rawsha256_fmt_plug.c
Comment (solardiz) [inline src/opencl_rawsha256_fmt_plug.c]: I think these comments don't help, and should be dropped. It is implied that if we're freeing resources, we're doing so not to leak them.

**Review by solardiz — APPROVED** [review]
I made some nitpick comments - feel free to make changes per them or not.

**Comment by magnumripper** [comment]
I added a commit fixing all other formats (bar a few strange Sayantan ones I don't grok). Unfortunetely I added that comment to all of them, lol.

File: src/opencl_rawsha256_fmt_plug.c
Comment (magnumripper) [reply src/opencl_rawsha256_fmt_plug.c:169]: What IDE, Eclipse?

Comment (ghost) [reply src/opencl_rawsha256_fmt_plug.c:169]: Visual Studio Code

**Comment by solardiz** [comment]
> Unfortunetely I added that comment to all of them, lol.

I don't understand this remark. Are you saying some formats have duplicate logic after that commit - e.g., both Claudio's changes and yours?

**Comment by magnumripper** [comment]
> > Unfortunetely I added that comment to all of them, lol.
> 
> I don't understand this remark. Are you saying some formats have duplicate logic after that commit - e.g., both Claudio's changes and yours?

No I added the redundant `// Avoid memory leaks` comment - but it's gone now due to some perl-fu and a force-push

**Review by solardiz — APPROVED** [review]
The 3 commits' combined changes look good to me now.

**Comment by magnumripper** [comment]
I'm happy with this now. No leaks seen, even in RAR 😵

**Comment by ghost** [comment]
I'll check it later.
```
Testing: o5logon-opencl, Oracle O5LOGON protocol [SHA1 AES OpenCL]... *** Error in `../run/john': malloc(): memory corruption: 0x00007f83929b5710 ***
```
```
Testing: o5logon-opencl, Oracle O5LOGON protocol [SHA1 AES OpenCL]... 0: OpenCL CL_INVALID_MEM_OBJECT (-38) error in opencl_o5logon_fmt_plug.c:143 - Error Unmapping result
```

**Comment by solardiz** [comment]
@claudioandre-br Are these o5logon-opencl failures new with this PR's changes or did they occur before? What system/device are you seeing them on?

**Comment by solardiz** [comment]
Oh, I see these are on Travis. So it's something we must look into before merging this.

**Comment by ghost** [comment]
We can squash commits:
2a9e64fb8b29fd3ed65d71b6fb774e96a83464de
56e9bcdea7956591e5d9930f6ff12f2ead8960d0
ba4fdba843b32dfd7ab70b66b1adc469f16829f6

If you guys think this is mandatory, I will create a new PR (because I need to keep this branch to compare).

**Comment by solardiz** [comment]
Now Travis failed on:

```
Testing: raw-MD4-opencl [MD4 OpenCL]... 0: OpenCL CL_INVALID_VALUE (-30) error in opencl_hash_check_128_plug.c:374 - failed in clEnqueueWriteBuffer buffer_bitmaps.
```

After fixing whatever other issues there might be, I think we should squash all commits here (currently all 6 of them). We can do that via GitHub UI if we don't mind GitHub appearing as the committer in git history, or @claudioandre-br can create a new PR for that as he suggested.

**Comment by ghost** [comment]
IMO, these commits are not generic and should be maintained.
c6ae3795ec1e5c09bd57126e570e718eccdc6c07 
1832a5b5a9f74bd38a4d6314be2fb3bd85c6e9fa

**Comment by magnumripper** [comment]
I see no particular reason to squash at all

**Comment by magnumripper** [comment]
...but I'm not merging this now if that's a concern, please do at will

File: src/opencl_rawsha1_fmt_plug.c
Comment (magnumripper) [inline src/opencl_rawsha1_fmt_plug.c:195]: I'd have set them as NULL, not that it makes any difference

**Review by magnumripper — APPROVED** [review]

**Comment by solardiz** [comment]
> IMO, these commits are not generic and should be maintained.
> [c6ae379](https://github.com/openwall/john/commit/c6ae3795ec1e5c09bd57126e570e718eccdc6c07)
> [1832a5b](https://github.com/openwall/john/commit/1832a5b5a9f74bd38a4d6314be2fb3bd85c6e9fa)

As I understand, the first one of these not only fixes but also introduces now-known bugs, and the second one fixes those. I'd squash.

File: src/opencl_rawsha1_fmt_plug.c
Comment (solardiz) [reply src/opencl_rawsha1_fmt_plug.c:195]: Sure, let's use NULL for pointers.

**Comment by solardiz** [comment]
Claudio, if you don't want the 6 commits squashed into one, please create a new PR with clean commits where we don't have a commit introduce known bugs nor even comments that we end up deleting later in the PR. Each commit should be reasonable on its own (at least not introducing new known bugs). While at it, please also take care of magnum's comment about NULLs. Thanks!

**Comment by ghost** [comment]
I tested (up to 100 iterations):
- rawMd5-OpenCL (still leaking; probably, the author of the format missed some buffer, maybe program[], queue[], ...
- rawSHA256-OpenCL (Ok)
- rawSHA512-OpenCL (Ok)
- xSHA512-OpenCL (Ok)
- SHA256Crypt-OpenCL (Ok)
- SHA512Crypt-OpenCL (Ok)
****
BTW: I'm testing on NVIDIA 1080. Is it safe, right? Or it'll burn?

**Comment by solardiz** [comment]
> BTW: I'm testing on NVIDIA 1080. Is it safe, right? Or it'll burn?

It's fine, but I suggest moving to the Titan X.

**Comment by ghost** [comment]
> It's fine, but I suggest moving to the Titan X.

I like to run these tests on 1080 (how the leak evolves). But I will at least let it rest.

**Comment by ghost** [comment]
Final thoughts: all formats need to be audited (later). #4295 should be kept open.

**Comment by solardiz** [comment]
> I will at least let it rest.

This is unnecessary. It's just that you get a severely throttled clock rate after only a few minutes on this poor card. If you only care about memory and not speeds, and you need to compare against the reference memory usage figures you saw earlier, you can simply continue to use the card like you've been using it.

**Comment by magnumripper** [comment]
I've tried to find the leak in raw-md5-opencl. Beats me so far.

Notes
After first test round, 168 MiB is lost. That's after a GWS of 16384. Then after each more round w/ same size, another 157 MiB is lost.

**Comment by magnumripper** [comment]
OK so I used this code with `--stress-test` option:
```diff
diff --git a/src/bench.c b/src/bench.c
index fa3191976..20a7c9262 100644
--- a/src/bench.c
+++ b/src/bench.c
@@ -732,6 +732,9 @@ int benchmark_all(void)
 		puts("NOTE: This is a debug build, speed will be lower than normal");
 #endif
 
+	fprintf(stderr, "GPU memory at start: ");
+	if (system("nvidia-smi --query-gpu=memory.used --format=csv,noheader"))
+		error();
 AGAIN:
 	options.loader.field_sep_char = 31;
 #endif
@@ -1110,6 +1114,8 @@ next:
 		}
 #ifndef BENCH_BUILD
 	if (options.flags & FLG_LOOPTEST) {
+		if (system("nvidia-smi --query-gpu=memory.used --format=csv,noheader"))
+			error();
 		if (event_abort) {
 			uint32_t p = 100 * loop_fail / loop_total;
 			uint32_t pp = 10000 * loop_fail / loop_total - p * 100;
```

And these debug macros:
```diff
diff --git a/src/opencl_common.h b/src/opencl_common.h
index 727f91731..76b610caf 100644
--- a/src/opencl_common.h
+++ b/src/opencl_common.h
@@ -287,6 +287,16 @@ void opencl_process_event(void);
 		} \
 	} while (0)
 
+/* Debug functions */
+#define jtrCreateBuffer(buf, context, flags, size, wat, ret)	  \
+	do { \
+		if ((buf)) \
+			error_msg("** %s:%d Memory leak\n", __FILE__, __LINE__); \
+		(buf) = clCreateBuffer((context), (flags), (size), (wat), (ret)); \
+	} while (0)
+
+#define jtrReleaseMemObject(buf) { cl_mem _b = (buf); (buf) = NULL; clReleaseMemObject(_b); }
+
 /* Macro for get a multiple of a given value */
 #define GET_NEXT_MULTIPLE(dividend, divisor)	  \
 	(divisor) ? ((dividend + (ocl_v_width * divisor - 1)) / (ocl_v_width * divisor)) * divisor : (dividend) / ocl_v_width;
```

I replaced ALL instances of creating or releasing buffers in opencl_rawmd5_fmt_plug.c and opencl_hash_check_128_plug.c - there simply is no leak there. I verified that if I commented out some clReleaseMemObject line, my macros catched it.

Still:
```
GPU memory at start: 205 MiB
Device 1: GeForce RTX 2080 Ti
#1 Benchmarking: raw-MD5-opencl [MD5 OpenCL/mask accel]... LWS=32 GWS=16384 DONE
Raw:    7166M c/s real, 7237M c/s virtual, Dev#1 util: 100%

373 MiB
#2 Benchmarking: raw-MD5-opencl [MD5 OpenCL/mask accel]... LWS=64 GWS=32768 DONE
Raw:    14333M c/s real, 14333M c/s virtual, Dev#1 util: 99%

530 MiB
#3 Benchmarking: raw-MD5-opencl [MD5 OpenCL/mask accel]... LWS=64 GWS=32768 DONE
Raw:    14333M c/s real, 14263M c/s virtual, Dev#1 util: 99%

687 MiB
#4 Benchmarking: raw-MD5-opencl [MD5 OpenCL/mask accel]... LWS=64 GWS=32768 DONE
Raw:    14333M c/s real, 14263M c/s virtual, Dev#1 util: 99%

844 MiB
#5 Benchmarking: raw-MD5-opencl [MD5 OpenCL/mask accel]... LWS=64 GWS=32768 DONE
Raw:    14254M c/s real, 14183M c/s virtual, Dev#1 util: 99%

1001 MiB
```

Obviously I've also had a good look at alloc/calloc/free (EDIT I now added assertions for all of them, no problem seen). Also, the format appears to destroy mappings, kernel and program correctly but maybe I should write more debug assertions like the above one in opencl_common.h.

EDIT 2: Same leak seen without auto-tune.

EDIT 3: Same leak seen with `--test --no-mask`

**Comment by magnumripper** [comment]
Just self-test without benchmark does not leak!
```
$ ../run/john --stress-test=0 -form:raw-md5-opencl 
GPU memory at start: 205 MiB
Device 1: GeForce RTX 2080 Ti
#1 Testing: raw-MD5-opencl [MD5 OpenCL]... PASS
216 MiB
#2 Testing: raw-MD5-opencl [MD5 OpenCL]... PASS
216 MiB
#3 Testing: raw-MD5-opencl [MD5 OpenCL]... PASS
216 MiB
```
Same for higher LWS/GWS (I'm not certain it's honored though, it used to be)
```
$ ../run/john --stress-test=0 -form:raw-md5-opencl -lws=64 -gws=4194304
GPU memory at start: 205 MiB
Device 1: GeForce RTX 2080 Ti
#1 Testing: raw-MD5-opencl [MD5 OpenCL]... PASS
216 MiB
#2 Testing: raw-MD5-opencl [MD5 OpenCL]... PASS
216 MiB
#3 Testing: raw-MD5-opencl [MD5 OpenCL]... PASS
216 MiB
```
EDIT: The higher GWS was **not** honored.

**Comment by ghost** [comment]
> OK so I used this code with --stress-test option:

Me likes some debug code seen here. Very useful.

**Comment by magnumripper** [comment]
> Me likes some debug code seen here. Very useful.

I added them in a separate commit - I think we should keep them.

**Comment by magnumripper** [comment]
Running those debug functions and `--format=opencl`, it looks like we're leaking a little even though I don't get any detected misses in buffer releases. We should probably add similar debug macros for `cl(Create|Release)Kernel()` and `cl(Create|Release)Program()`.

EDIT: Did that. Now it looks more like GPU memory is "slow" to recover, or fragmented. Any given single format will not "leak" while running all of them will look like they do, a little. Note also that many formats use `mem_alloc_tiny` somewhere which in itself is a memory leak when running this way - but usually not IRL.

**Comment by magnumripper** [comment]
DEScrypt-opencl doesn't even have a `done()` 👀

**Comment by magnumripper** [comment]
I get segfaults from mscash2-opencl, disabling it from more tests. Sayantan again, it's a pity he's not with us now

**Comment by magnumripper** [comment]
Got this now
```
#2 Testing: raw-SHA256-opencl [SHA256 OpenCL]... 0: OpenCL CL_INVALID_KERNEL (-48) error in opencl_rawsha256_fmt_plug.c:150 - Error setting argument 6
```
It's reproducible alone as `../run/john --stress-test=0 -form:raw-sha256-opencl`. My debugging stuff was in place. EDIT: It doesn't have to be.

**Comment by magnumripper** [comment]
> DEScrypt-opencl doesn't even have a `done()` 👀

I'm opening a separate issue for that one. It's not a trivial one to fix, I always hated trying to follow the strange code paths through different files.

**Comment by magnumripper** [comment]
After 35dd7fc, I can run all formats (except mscash2-opencl that just segfaults and descrypt-opencl which has #4362, I refuse to look at them) several times with no detected leak. I think this PR is ready to merge.

**Comment by magnumripper** [comment]
> After [35dd7fc](https://github.com/openwall/john/commit/35dd7fc22d31294e0118e06202d46039bff51327), I can run all formats (except mscash2-opencl that just segfaults and descrypt-opencl which has #4362, I refuse to look at them) several times with no detected leak. I think this PR is ready to merge.

No wait, I only ran through `--stress-test=0` several rounds. I obviously need to do a `--stress-test` a couple rounds also.

**Comment by ghost** [comment]
> except mscash2-opencl [...] and descrypt-opencl 

To test, we need to remove hard to maintain formats. :+1:

**Comment by magnumripper** [comment]
I see now many formats have this helper macro
```
#define CLCREATEBUFFER(_flags, _size, _string) \
   clCreateBuffer(context[gpu_id], _flags, _size, NULL, &cl_error); \
   HANDLE_CLERROR(cl_error, _string);
```
My previous perl-fu for patching in the debug functions would miss them

**Comment by solardiz** [comment]
> except mscash2-opencl that just segfaults

@magnumripper Please create an issue for the mscash2-opencl segfault, including reproduction instructions. Thanks!

**Comment by solardiz** [comment]
> To test, we need to remove hard to maintain formats. +1

But we need to keep those formats available for use. I think mscash2-opencl is fairly easy to rewrite in a manner consistent with other formats, and that should be done (we'll lose this format's unique ability/hack to use multiple devices without forking, but that's OK). descrypt-opencl is much harder.

**Comment by magnumripper** [comment]
I just add them to disabled formats section in my `john-local.conf`, no problem.

**Comment by solardiz** [comment]
> I think this PR is ready to merge.

I don't like having the 13 separate commits. I think the separation between some of them is more historical (during the 3-day history of this PR) rather than logical. Ideally, you'd squash some of the commits manually, having fewer properly logically separated commits to merge, and with titles reflecting their differences from other commits here. Would you do that, @magnumripper? Thanks!

File: src/bench.c
Comment (solardiz) [inline src/bench.c:14]: Hmm. I've just confirmed there are no uses of `setenv` in `bench.c`. OK.

File: src/bench.c
Comment (solardiz) [inline src/bench.c]: Maybe we can limit this to OpenCL formats? Ideally, also to uses of an NVIDIA device, but that's harder. Also, please add braces around `options.flags & FLG_LOOPTEST`.

File: src/bench.c
Comment (solardiz) [inline src/bench.c]: Should probably be inside `#ifndef BENCH_BUILD`.

File: src/bench.c
Comment (solardiz) [inline src/bench.c:1080]: Should probably be inside `#ifndef BENCH_BUILD`. And please test `make generic`.

File: src/opencl_common.h
Comment (solardiz) [inline src/opencl_common.h:315]: I think we should standardize on `john` in naming of related projects, identifiers in code, etc. Not `jtr`.

File: src/opencl_rawsha1_fmt_plug.c
Comment (solardiz) [reply src/opencl_rawsha1_fmt_plug.c:195]: I see we still use 0 and not NULL here. These are `cl_mem`, so it's non-obvious whether any such usage is correct at all - maybe that was meant to be an opaque type for which we don't reliably know whether it's an integer, pointer, or something else? On a given system I see it's `typedef struct _cl_mem *            cl_mem;`, so maybe NULL is more correct, but ideally we'd know more about what's guaranteed (or not) about this type and use these variables only as guaranteed to work by OpenCL spec.

**Review by solardiz — APPROVED** [review]
I'm posting some review comments, even though I see you've already pushed more commits that I haven't included yet.

**Comment by magnumripper** [comment]
> > I think this PR is ready to merge.

That was premature, still testing.

> I don't like having the 13 separate commits. I think the separation between some of them is more historical (during the 3-day history of this PR) rather than logical. Ideally, you'd squash some of the commits manually, having fewer properly logically separated commits to merge, and with titles reflecting their differences from other commits here. Would you do that, @magnumripper? Thanks!

Will do. I use git autosquash feature so my recent (strange) commit descriptions already tell git how to handle half of it when I later do a `git rebase -i`.

File: src/bench.c
Comment (magnumripper) [reply src/bench.c:14]: Yeah I wondered where they went but didn't look further into it

File: src/opencl_common.h
Comment (magnumripper) [reply src/opencl_common.h:315]: Will try to remember that for #4363 (perhaps I'll end up using neither, just eg. OCLRELEASEMEM)

**Comment by magnumripper** [comment]
I've had occasional segfaults from mysql-sha1-opencl and mscash-opencl as well now, but can't reproduce. It really looks like Sayantan did something in his code that either isn't quite compatible with our shared code, or just plain wrong in some other aspect or detail - virtually all such problems are with his few formats. It's just that I've spent hours and hours trying to find it over the years and I can't see any particular problem! I bet once we find it, it's a similar problem across all of them.

**Comment by magnumripper** [comment]
I see things like notes-opencl segfaulting on an all-formats test (right after LM-opencl, who would have thought) but not on its own (we've seen that before). I'm actually considering adding option `--format-list=LIST` so I could test with `--format=lm-opencl,notes-opencl`. I'd use the existing list parsing of course, like we do with several other recent-ish options like `--single-seed=WORD[,WORD...]`

**Comment by magnumripper** [comment]
Done testing now. I'll merge once bots are happy

**Pull request merged**
